Log with attached image returns upload failure although successful

### Describe your problem!

On support we have two new tickets stating, that on logging with an attached picture c:geo returns an error, even on several retries. But when they check online with their browser, the log got sent.

Both tickets are about logs with attached images, don't know yet whether the error message also occurs if no image is attached (but I sort of doubt it, otherwise we would have tons of support tickets).

Logcat does not show any related errors for them.


I tried to reproduce this from within the emulator, and I'm not able to send a log at all, I always get an "unknown error". Logcat lists this:
```
2022-02-04 21:45:15.791 17728-17975/cgeo.geocaching E/cgeo: [AsyncTask #2] Error posting log
    java.lang.RuntimeException: java.io.IOException: request was not successful: Response{protocol=http/1.1, code=400, message=Bad Request, url=https://www.geocaching.com/api/proxy/web/v1/geocache/gc12345/GeocacheLog}
        at io.reactivex.rxjava3.internal.util.ExceptionHelper.wrapOrThrow(ExceptionHelper.java:46)
        at io.reactivex.rxjava3.internal.observers.BlockingMultiObserver.blockingGet(BlockingMultiObserver.java:94)
        at io.reactivex.rxjava3.core.Single.blockingGet(Single.java:3645)
        at cgeo.geocaching.connector.gc.GCWebAPI.postLog(GCWebAPI.java:926)
        at cgeo.geocaching.connector.gc.GCLoggingManager.postLog(GCLoggingManager.java:137)
        at cgeo.geocaching.log.LogCacheActivity$Poster.doInBackgroundInternal(LogCacheActivity.java:687)
        at cgeo.geocaching.log.LogCacheActivity$Poster.doInBackgroundInternal(LogCacheActivity.java:671)
        at cgeo.geocaching.utils.AbstractAsyncTaskWithProgress.doInBackground(AbstractAsyncTaskWithProgress.java:134)
        at android.os.AsyncTask$3.call(AsyncTask.java:378)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.io.IOException: request was not successful: Response{protocol=http/1.1, code=400, message=Bad Request, url=https://www.geocaching.com/api/proxy/web/v1/geocache/gc12345/GeocacheLog}
        at cgeo.geocaching.network.Network.lambda$static$6(Network.java:654)
        at cgeo.geocaching.network.-$$Lambda$Network$HJwZsCLFMmiISrdxtajFdFJJscU.apply(Unknown Source:2)
        at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:77)
        at io.reactivex.rxjava3.internal.operators.single.SingleCreate$Emitter.onSuccess(SingleCreate.java:68)
        at cgeo.geocaching.utils.RxOkHttpUtils$1.onResponse(RxOkHttpUtils.java:48)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
```

(Not sure if those two errors are related, or if the cause for the failing call in the emulator is something emulator-specific.)

### How to reproduce?

- open c:geo
- open a cache
- send log, with or without attached image

### Actual result after these steps?

- in my emulator test:
  - a popup opens with "unknown error, see FAQ" (etc.)
  - log did not get sent
- for the users on support
  - some error is being displayed
  - but the log got sent


### Expected result after these steps?

- generally: post log successfully
- show error message only when an actual error has occurred


### Reproducible

Yes

### c:geo Version

2021.12.24

### System information

_No response_

### Additional Information

Tickets 876615 and 417247

--- a/main/src/cgeo/geocaching/connector/gc/GCLoggingManager.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCLoggingManager.java
@@ -172,7 +172,7 @@ class GCLoggingManager extends AbstractLoggingManager implements LoaderManager.L
 
         if (!image.isEmpty()) {
 
-            final ImmutablePair<StatusCode, String> imageResult = GCWebAPI.postLogImage(logId, image);
+            final ImmutablePair<StatusCode, String> imageResult = GCWebAPI.postLogImage(cache.getGeocode(), logId, image);
 
             return new ImageResult(imageResult.left, imageResult.right);
         }

--- a/main/src/cgeo/geocaching/connector/gc/GCWebAPI.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCWebAPI.java
@@ -20,8 +20,10 @@ import cgeo.geocaching.network.Parameters;
 import cgeo.geocaching.utils.AndroidRxUtils;
 import cgeo.geocaching.utils.CollectionStream;
 import cgeo.geocaching.utils.Log;
+import cgeo.geocaching.utils.TextUtils;
 
 import androidx.annotation.NonNull;
+import androidx.core.util.Consumer;
 
 import java.io.File;
 import java.text.DateFormat;
@@ -36,6 +38,7 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Set;
 import java.util.UUID;
+import java.util.regex.Pattern;
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 import com.fasterxml.jackson.annotation.JsonProperty;
@@ -58,7 +61,12 @@ import org.apache.commons.lang3.tuple.ImmutablePair;
 class GCWebAPI {
 
     private static final Object CACHE_LOCK = new Object();
-    private static final String API_URL = "https://www.geocaching.com/api/proxy";
+    private static final String WEBSITE_URL = "https://www.geocaching.com";
+    private static final String API_URL = WEBSITE_URL + "/api/proxy";
+
+    //<input name="__RequestVerificationToken" type="hidden" value="(token-value)" />
+    private static final Pattern PATTERN_REQUEST_VERIFICATION_TOKEN = Pattern.compile("name=\"__RequestVerificationToken\"\\s+type=\"hidden\"\\s+value=\"([^\"]+)\"");
+
 
     /** maximum number of elements to retrieve with one call */
     private static final int MAX_TAKE = 50;
@@ -774,8 +782,13 @@ class GCWebAPI {
         return getAuthorizationHeader().flatMap((Function<Parameters, SingleSource<T>>) headers -> Network.postRequest(API_URL + path, clazz, parameters, headers, fileFieldName, fileContentType, file).subscribeOn(AndroidRxUtils.networkScheduler));
     }
 
-    private static Single<Response> postAPI(final String path, final Object jsonObject) {
-        return getAuthorizationHeader().flatMap((Function<Parameters, Single<Response>>) headers -> Network.postJsonRequest(API_URL + path, headers, jsonObject).subscribeOn(AndroidRxUtils.networkScheduler));
+    private static Single<Response> postAPI(final String path, final Object jsonObject, final Consumer<Parameters> headerAdder) {
+        return getAuthorizationHeader().flatMap((Function<Parameters, Single<Response>>) headers -> {
+            if (headerAdder != null) {
+                headerAdder.accept(headers);
+            }
+            return Network.postJsonRequest(API_URL + path, headers, jsonObject).subscribeOn(AndroidRxUtils.networkScheduler);
+        });
     }
 
     static Single<CacheDetails> getCacheDetails(final String geocode) {
@@ -995,7 +1008,7 @@ class GCWebAPI {
     }
 
     private static Response postLogTrackable(final List<TrackableLog> trackableLogs) {
-        final Response response = postAPI("/trackable/activities", trackableLogs).blockingGet();
+        final Response response = postAPI("/trackable/activities", trackableLogs, (Consumer<Parameters>) null).blockingGet();
         if (!response.isSuccessful()) {
             Log.e("Logging trackables failed: " + response.message());
         }
@@ -1028,6 +1041,9 @@ class GCWebAPI {
      * Post an image and attach it to the log.
      *
      * The following sequence of http requests have to be performed:
+     *
+     * 0) GET to https://www.geocaching.com/play/geocache/(geocode)/log
+     * Grab value of input field "__RequestVerificationToken" from HTML, is needed for request 3) below
      *
      * 1)
      * Post image to: https://www.geocaching.com/api/proxy/web/v1/LogDrafts/images
@@ -1066,12 +1082,25 @@ class GCWebAPI {
      * 3)
      * POST: https://www.geocaching.com/api/proxy/web/v1/geocaches/logs/GL.../images/14242d4d-...
      * Request:
+     * Header: X-Verification-Token: (verificationToken)
+     * Body:
      * <pre>
      *     {"name":"","uuid":"","guid":"14242d4d-ca1f-425e-9496-aa830d769350","thumbnailUrl":"https://img.geocaching.com/large/14242d4d-...jpg","dateTaken":"2017-09-07","description":"","qqDropTarget":{},"id":3,"filename":"filename.png","lastModified":1494143750916,"lastModifiedDate":"2017-05-07T07:55:50.916Z","webkitRelativePath":"","size":13959,"type":"image/png"}
      * </pre>
      * Response not used
      */
-    static ImmutablePair<StatusCode, String> postLogImage(final String logId, final Image image) {
+    static ImmutablePair<StatusCode, String> postLogImage(final String geocode, final String logId, final Image image) {
+
+        // 0) open log page to get a Request Token
+        final String html = Network.getResponseData(Network.getRequest(WEBSITE_URL + "/play/geocache/" + geocode + "/log"));
+        if (html == null) {
+            return new ImmutablePair<>(StatusCode.LOGIMAGE_POST_ERROR, null);
+        }
+        final String requestVerificationToken = TextUtils.getMatch(html, PATTERN_REQUEST_VERIFICATION_TOKEN, null);
+        if (requestVerificationToken == null) {
+            return new ImmutablePair<>(StatusCode.LOGIMAGE_POST_ERROR, null);
+        }
+
         // 1) upload image to drafts
         final Parameters params = new Parameters();
         params.put("guid", UUID.randomUUID().toString());
@@ -1094,7 +1123,9 @@ class GCWebAPI {
         attachImageRequest.thumbnailUrl = postImageResponse.thumbnailUrl;
         attachImageRequest.name = StringUtils.defaultString(image.getTitle());
         attachImageRequest.description = StringUtils.defaultString(image.getDescription());
-        final Response attachResponse = postAPI("/web/v1/geocaches/logs/" + logId + "/images/" + postImageResponse.guid, attachImageRequest).blockingGet();
+        final Response attachResponse = postAPI("/web/v1/geocaches/logs/" + logId + "/images/" + postImageResponse.guid, attachImageRequest, h -> {
+            h.add("X-Verification-Token", requestVerificationToken);
+        }).blockingGet();
         if (!attachResponse.isSuccessful()) {
             return new ImmutablePair<>(StatusCode.LOGIMAGE_POST_ERROR, null);
         }

